### PR TITLE
add prefer_null_aware_method_calls

### PR DESCRIPTION
**Pull request by a14n**
# Description

Instead of checking nullability of a function/method `f` before calling it you
can use `f?.call()`.

**BAD:**

```
if (f != null) f!();
```

**GOOD:**

```
f?.call();
```

Requested by https://github.com/flutter/flutter/pull/79003#issuecomment-806156141

### CODE REVIEW

File: lib/src/rules/prefer_null_aware_method_calls.dart [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+  _Visitor(this.rule);
+
+  @override
+  void visitIfStatement(IfStatement node) {
+    var condition = node.condition;
+    if (condition is BinaryExpression &&
+        condition.operator.type == TokenType.BANG_EQ &&
+        condition.rightOperand is NullLiteral &&
+        node.elseKeyword == null) {
```
Comment (bwilkerson) [inline lib/src/rules/prefer_null_aware_method_calls.dart:59]: Neither of these conditions check to see that the left hand side is a reference to a function, nor do they check that the function being invoked is the same as the one being tested. I think it's safe to ignore the first check as long as we check the second condition. For example, let's add a test of the form
```
var f, g;
void m() {
  if (f != null) g!();
  f != null ? g!() : null;
}
```

Comment (a14n) [reply lib/src/rules/prefer_null_aware_method_calls.dart:59]: The equality between condition and target of call is done at line 89. The test you mentioned is similar to lines 19 and 25 of the test file.

Comment (bwilkerson) [reply lib/src/rules/prefer_null_aware_method_calls.dart:59]: Sorry, I missed that.

File: lib/src/rules/prefer_null_aware_method_calls.dart [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+      _check(node.thenExpression, condition);
+    }
+  }
+
+  void _check(Expression expression, BinaryExpression condition) {
+    if (expression is FunctionExpressionInvocation) {
+      var target = expression.function;
+      if (target is PostfixExpression &&
+          target.operator.type == TokenType.BANG &&
+          target.operand.toString() == condition.leftOperand.toString()) {
```
Comment (bwilkerson) [inline lib/src/rules/prefer_null_aware_method_calls.dart:89]: The `analyzer` APIs reserve `toString` for debugging purposes. Technically these should be converted to use `toSource`.

**Review by bwilkerson — APPROVED** [review]

File: lib/src/rules/prefer_null_aware_method_calls.dart [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/token.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+
+import '../analyzer.dart';
+
+const _desc = r'Prefer null aware method call.';
```
Comment (pq) [inline lib/src/rules/prefer_null_aware_method_calls.dart:11]: =>

    r'Prefer null aware method calls.'

?

**Review by pq — APPROVED** [review]
👍

**Pull request merged**